### PR TITLE
Radio Detection Tool

### PR DESCRIPTION
**Pull request by Super3222**
So like I mentioned in https://github.com/yogstation13/yogstation/pull/307 I was planning on making another PR that had almost the same ability as GPS Scouters.

Introducing... Radio Pulsers/Radio Detection Tools/Radio Detection Devices!
They basically allow you to send out a signal to people specifically wearing headsets, that returns back with information like which type of headset they have on, which channels are associated with it, and how far away is it.

It costs only 5 TC (from 3 TC)!

Their scan radius can be tweaked by using CTRL + CLICK to improve it and ALT + CLICK to lower it. There will be an 'intensity' rating which grows smaller or larger when you tweak it's values.

By default, the intensity rating will start at 3.

For reference on how you want to tweak your pulser, I was able to pick up the QM's signal (he was in maintenance) from the wardens desk while using an intensity of 3.

This would be a pretty neat combination with the default night vision goggles Nuke Ops. Since, not many of them want to lose their nightvision for optical mesons.

Here's some convenient stats:
Intensity | Scan Radius
5 | 35
4 | 30
3 | 25
2 | 20
1 | 15

Screenshots:

Also, for godsake ignore that middle commit message.

:cl:Super3222
rscadd: There is a new syndicate tool avaliable for nuclear operatives called a radio detection device.
rscadd: It's ability is to pick up signals from other radios. It's intensity can be tweaked to manipulate the scan radius to the user's liking.
rscadd: The radio detection device can also be manipulated to only focus on certain types of radios!
rscadd: It is available for 1 TC!
/:cl:

### CODE REVIEW

**Comment by ghost** [comment]
this is not 5tc item

this is a 2 or 1 tc item

**Comment by ShadowDeath6** [comment]
still don't think ops need this, it's better for traitors

**Comment by Super3222** [comment]
More tweaks and adjustments going to be made. Stay tuned and don't forget to leaf a like and subscribe

**Comment by Super3222** [comment]
In the last commit:
- Fixes a bug which prevented CtrlShiftClickOn and ShiftAltClickOn from allowing humans to use on objects.
- The Pulser can now be focused on channels like:


- Cost is now 1 TC.

File: code/game/objects/items/devices/radiopulser.dm [hunk 1]
```diff
@@ -0,0 +1,196 @@
+#define PULSECOOLDOWN 350
+
+/obj/item/device/pulse
+	name = "radio pulser"
+	desc = "With one click of a button, this device will send a 360 electromagnetic burst which is picked up by various \
+			headset devices associated with a heat signature."
```
Comment (Alek2ander) [inline code/game/objects/items/devices/radiopulser.dm:6]: > 360 ~~electromagnetic burst~~ noscope

360 degree at least.

> headset devices associated with a heat signature.

Wut

File: code/game/objects/items/devices/radiopulser.dm [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+	var/cooldown
+	var/scanning = FALSE
+	var/intensity = 3 // max 35, min 15. for people who want longer or closer ranges scanned
+	var/scanradius = 25 // default scan radius
+	var/freqhunt // none equals universal
+
+
+/obj/item/device/pulse/examine(mob/user)
+	..()
+	user << "<span class='notice'>The intensity meter is aiming at a [intensity].</span>"
```
Comment (Alek2ander) [inline code/game/objects/items/devices/radiopulser.dm:25]: pointing to

File: code/game/objects/items/devices/radiopulser.dm [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+/obj/item/device/pulse/proc/gettargetrange(mob/user, obj/item/device/radio/headset/target_radio)
+	var/shortrange = scanradius / 5
+	var/midrange = scanradius / 2
+	var/turf/T = get_turf(user)
+	var/turf/TR = get_turf(target_radio)
+
+	if(T.Distance(TR) <= shortrange)
+		return "You are at a short distance away from the nearby radio!"
+	if(T.Distance(TR) <= midrange)
+		return "You are at a midrange distance away from the radio!"
```
Comment (Alek2ander) [inline code/game/objects/items/devices/radiopulser.dm:194]: medium distance

Also, direction would be nice

File: code/game/objects/items/devices/radiopulser.dm [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+				keyslots++
+				for(var/A in attracted_radio.keyslot.channels)
+					user << "<span class='notice'>[A] channel frequencies are located within the radio!</span>"
+					if(prob(5))
+						H << "<span class='danger'>[attracted_radio] begins buzzing.</span>"
+
+			if(attracted_radio.keyslot2)
+				keyslots++
+				for(var/A in attracted_radio.keyslot2.channels)
+					user << "<span class='notice'>[A] channel is located within the radio!</span>"
```
Comment (Alek2ander) [inline code/game/objects/items/devices/radiopulser.dm:77]: Different message from keyslot 1, and no chance of feedback

File: code/game/objects/items/devices/radiopulser.dm [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+		return
+	else
+		scanning = TRUE
+	user << "<span class='danger'>You trigger the pulser! (Activate the pulser again to disable it's scanning mode)</span>"
+	var/turf/T = get_turf(src)
+	var/list/scanned_radios = list()
+	var/reached // whether the signal reached a valid headset
+
+
+	for(var/mob/living/carbon/human/H in orange(scanradius,T))
```
Comment (Alek2ander) [inline code/game/objects/items/devices/radiopulser.dm:48]: The worst way to iterate this. A better way to be by moblist or by radio lists.

File: code/game/objects/items/devices/radiopulser.dm [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+			if(dat)
+				scanned_radios.Add(attracted_radio)
+				if(!reached)
+					reached = 1
+			user << "<span class='notice'>*-------*</span>"
+			if(!scanning)
+				break
+			user << "<span class='alert'>Attempting to locate more radios . . .</span>"
+			sleep(30)
+			user << "<span class='alert'>Locating another signal . . .</span>"
```
Comment (Alek2ander) [inline code/game/objects/items/devices/radiopulser.dm:94]: user might not exist

**Comment by Alek2ander** [comment]
Other problems:
Will detect disabled radios
Will _not_ detect station-bounced radios, or radios not being worn on ears.
Will spam your chat with dozens of pointless fluff lines.

**Comment by Super3222** [comment]
I'm going into massively buffing this in the future and get rid of the fluff lines (and other errors) but, for now I'd rather not have a stale PR up.

**Comment by ShadowDeath6** [comment]
rip

**Pull request closed without merging**